Overwrite restrictions documentation for GCP (#25645)
hashicorp/vault: website/content/docs/sync/gcpsm.mdx

default credentials. The service account must be [authorized](https://cloud.google.com/iam/docs/service-account-overview#locations)
to perform Secret Manager actions in the target project.

## Security

Access to secrets can be allowed or restricted through the use of their
[resource names by using conditions](https://cloud.google.com/iam/docs/conditions-resource-attributes#resource-name). For secrets,
the fully qualified resource name is required in the format of `projects/<project_number>/secrets/<secret_name>` Take note that

> this requires the project _number_, not the project _ID_.

Reviewer: Any reason you did not use the tag/label condition like for AWS?

Tags are more commonly used as conditions and they do not depend on user-defined fields like the name template making them less error prone. We should guide users to the solution with less pitfalls.

Author: From my research, GCP doesn't seem to support resource tags on individual secrets, and additionally it also doesn't support using secret labels as IAM conditions.

I tried finding alternatives but the only solution I was able to find was creating this rule via secret name. Are you already familiar with a better option? I did find feature requests from users for Google to add tags to secrets but in the meantime I don't believe there's other options.

Reviewer: My bad you're right, [GCP SM only has labels and labels cannot be used for IAM conditions](https://cloud.google.com/compute/docs/labeling-resources#labels_and_tags). I thought they'd have the same capabilities.

> Access to secrets can be allowed or restricted through the use of their
> [resource names by using conditions](https://cloud.google.com/iam/docs/conditions-resource-attributes#resource-name). For secrets,
> the fully qualified resource name is required in the format of `projects/<project_number>/secrets/<secret_name>` Take note that
> this requires the project _number_, not the project _ID_.

Reviewer 2: ```suggestion
You can allow or restrict access to secrets based on IAM
[conditions](https://cloud.google.com/iam/docs/conditions-resource-attributes#resource-name)
against the fully qualified resource name.

<Tip title="Use the project number not the project ID">

  For Vault secrets, a fully qualified resource name must have the format:

  `projects/<project_number>/secrets/<secret_name>`

  The project **number** is not the same as the project **ID**. Project numbers
  are **numeric** while project IDs are **alphanumeric**.

</Tip>
```

> For example, if the default secret name template is used, which prepends the word `vault` to the beginning of secret names,
> then following Role binding condition using `startsWith` will prevent Vault from being able to modify secrets that were not created
> by a sync operation:

Reviewer 2: ```suggestion
For example, the default secret name template prepends the word `vault` to the
beginning of secret names. To prevent Vault from modifying secrets that were not
created by a sync operation, you can use a role binding against the resource
name with the `startsWith` condition:
```

Style correction: write in active voice

  <CodeBlockConfig hideClipboard>

    resource.name.startsWith("projects/<project_number>/secrets/vault")

>   </CodeBlockConfig>

Reviewer: Can we give a complete working example? not being super familiar with GCP I don't know how to use this condition to configure the role that will be used by Vault.

> Likewise, is it recommended to add a negative condition on other write-access role bindings not used by Vault to secrets to prevent
> out-of-band overwrites. This is done simply by applying the inverse to the same condition:

Reviewer 2: ```suggestion
We recommend also adding a negative/inverse condition on the write-access role
bindings not used by Vault to secrets to prevent out-of-band overwrites:
```

  <CodeBlockConfig hideClipboard>

    !(resource.name.startsWith("projects/<project_number>/secrets/vault"))

  </CodeBlockConfig>

  <Note>
  Wildcards are used in the example above for brevity, but it is recommended to use the least privilege principle and restrict
  the actions and resources to the minimum requirements for each use case.

>   </Note>

Reviewer: I don't think this is relevant for GCP?

>   <Note>
>   Wildcards are used in the example above for brevity, but it is recommended to use the least privilege principle and restrict
>   the actions and resources to the minimum requirements for each use case.
>   </Note>

Reviewer 2: ```suggestion
<Warning title="Use wildcards with extreme caution">

  The previous examples use wildcards for the sake of brevity. We strongly
  recommend you use the principle of least privilege to restrict actions and
  resources for each use case to the minimum necessary requirements.
  
</Warning>
```

Updated to use the same language as the other document

</Tip>

For example, the default secret name template prepends the word `vault` to the
beginning of secret names. To prevent Vault from modifying secrets that were not
created by a sync operation, you can use a role binding against the resource
name with the `startsWith` condition:

  <CodeBlockConfig hideClipboard>

    resource.name.startsWith("projects/<project_number>/secrets/vault")

  </CodeBlockConfig>

To prevent out-of-band overwrites, simply add a negative condition with `!` on any
write-access role bindings not being used by Vault that contain Secret Manager permissions:

  <CodeBlockConfig hideClipboard>

    !(resource.name.startsWith("projects/<project_number>/secrets/vault"))

  </CodeBlockConfig>

To add these conditions and others to IAM principals in GCP, refer to the image below and the 
[IAM Conditions overview documentation](https://cloud.google.com/iam/docs/conditions-overview).

![GCP IAM Conditions](/img/gcp-add-iam-conditions_light.png#light-theme-only)
![GCP IAM Conditions](/img/gcp-add-iam-conditions_dark.png#dark-theme-only)

![GCP IAM Conditions](/img/gcp-add-iam-conditions_light2.png#light-theme-only)

> ![GCP IAM Conditions](/img/gcp-add-iam-conditions_dark2.png#dark-theme-only)

Author: Do you like the addition of this statement along with a screenshot showing where IAM Conditions can be added? @schavis 

If so, the circled version or uncircled version?

Reviewer 2: If we add an image, definitely use the circled one. But we need more useful alt text so it's meaningful for screen readers as well.
